File: algo/KthLargestElementinanArray.java
import java.util.Arrays;
import java.util.PriorityQueue;

class KthLargestElementinanArray {
    // 215. Kth Largest Element in an Array

    // Approach 1: PriorityQueue
    // Time Complexity: O(N)
    // Space Complexity: O(k)
    public int findKthLargest(int[] nums, int k) {
        PriorityQueue<Integer> qp = new PriorityQueue<>();
        for (int i = 0; i < nums.length; i++) {
            qp.add(nums[i]);
            if (qp.size() > k) {
                qp.remove();
            }
        }
        return qp.poll();
    }

    // Approach 3: Sort
    // Time Complexity: O(NlogN)
    // Space Complexity: O(1)
    public int findKthLargest2(int[] nums, int k) {
        Arrays.sort(nums);
        return nums[nums.length - k];
    }

}

File: algo/KthLargestElementinanArray.cpp
class Solution {
    
public:
    int findKthLargest(vector<int>& nums, int k) {
        k = nums.size()-k;
        return partition(nums, k, 0, nums.size()-1);
    }

private:
    int partition(vector<int>& nums, int k, int l, int r) {
        int pivot = nums[r];
        int p = 0;
        for (int i=0; i < r; i++) {
            if (nums[i] <= pivot) {
                swap(nums[i], nums[p++]);
            }
        }
        swap(nums[r], nums[p]);
        if (k > p) {
            return partition(nums, k, p+1, r);
        } else if (k < p) {
            return partition(nums, k, l, p-1);
        } else {
            return nums[p];
        }
    }    
};